src/Control/Distributed/Closure/Class.hs

```diff
@@ -0,0 +1,81 @@
+-- | 'Closure' is not a functor, since we cannot map arbitrary functions over
+-- it. But it sure looks like one, and an applicative one at that. What we can
+-- do is map /static pointers to/ arbitrary functions over it (or in general,
+-- closures). 'Closure' is not just an applicative functor, it's also a monad,
+-- as well as a comonad, if again we limit the function space to those functions
+-- that can be statically pointed to.
+--
+-- In fact an entire hierarchy of classes mirroring the standard classes can be
+-- defined, where nearly the only difference lies in the fact that higher-order
+-- arguments must be a proof of /static/-ness (i.e. a 'Closure'). The other
+-- difference is that composing static values requires a proof of typeability,
+-- so we carry those around ('Typeable' constraints).
+--
+-- This module defines just such a class hierarchy in the category of static
+-- functions (aka values of type @'Closure' (a -> b)@).
+
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE StaticPointers #-}
+{-# LANGUAGE UndecidableInstances #-} -- For StaticMonad instance.
+
+module Control.Distributed.Closure.Class
+  ( Static(..)
+  , StaticFunctor(..)
+  , StaticApply(..)
+  , StaticBind(..)
+  , StaticApplicative(..)
+  , StaticMonad
+  , staticReturn
+  , StaticComonad(..)
+  ) where
+
+import Control.Distributed.Closure
+import Data.Typeable (Typeable)
+
+class Typeable f => StaticFunctor f where
+  staticMap :: (Typeable a, Typeable b) => Closure (a -> b) -> f a -> f b
+
+class StaticFunctor f => StaticApply f where
+  staticApply :: (Typeable a, Typeable b) => f (a -> b) -> f a -> f b
+
+class StaticApply f => StaticApplicative f where
+  staticPure :: Typeable a => a -> f a
+
+class StaticApply m => StaticBind m where
```

Comment:
Ok. I don't think there needs to be a link `Apply => Bind` but let's keep the same relations a `semigroupoids`.

An also, now that I've thought a bit more, what we are actually defining here are endofunctors in the category of static functions. We've just massaged ignored some of the `Closure` for the sake of convenience. If we wrote everything as endofunctor, we would have, I think the following types:
```haskell
map :: Closure (a->b) -> Closure(f a -> f b)
pure :: Closure (a -> f a)
join :: Closure (f (f a) -> a)
bind :: Closure (a -> f b) -> Closure (f a -> f b)
ap :: Closure (f (a -> b) -> f a -> f b)
```

So I'm convinced with the types now. Maybe these are also of interest? Together with a `Category` instance on `Closure (a -> b)`? (we can also implement `(***)` if I'm not mistaken, but not `arr`, but maybe we should add a `StaticArrow` type-class as well?).